pennylane/ops/arithmetic/adjoint.py

```diff
@@ -0,0 +1,435 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule defines an operation modifier that indicates the adjoint of an operator.
+"""
+from functools import wraps
+
+from pennylane.operation import Operator, Operation, AnyWires, AdjointUndefinedError
+from pennylane.queuing import QueuingContext, QueuingError
+from pennylane.tape import QuantumTape, stop_recording
+from pennylane.math import transpose, conj
+
+
+# pylint: disable=too-many-public-methods
+class Adjoint(Operator):
+    """
+    The Adjoint of an operator.
+
+    Args:
+        base (~.operation.Operator): The operator that is adjointed.
+
+    .. seealso:: :func:`~.adjoint`, :meth:`~.operation.Operator.adjoint`
+
+    **Example:**
+
+    >>> op = Adjoint(qml.S(0))
+    >>> op.name
+    'Adjoint(S)'
+    >>> qml.matrix(op)
+    array([[1.-0.j, 0.-0.j],
+       [0.-0.j, 0.-1.j]])
+    >>> qml.generator(Adjoint(qml.RX(1.0, wires=0)))
+    (PauliX(wires=[0]), 0.5)
+    >>> Adjoint(qml.RX(1.234, wires=0)).data
+    [1.234]
+
+    """
+
+    num_wires = AnyWires
+
+    # pylint: disable=attribute-defined-outside-init
+    def __copy__(self):
+        # this method needs to be overwritten becuase the base must be copied too.
+        cls = self.__class__
+        copied_op = cls.__new__(cls)
+        copied_base = self.base.__copy__()
+        copied_op._hyperparameters = {"base": copied_base}
+        for attr, value in vars(self).items():
+            if attr not in {"data", "base", "_hyperparameters"}:
+                setattr(copied_op, attr, value)
+
+        return copied_op
+
+    # pylint: disable=super-init-not-called
+    def __init__(self, base=None, do_queue=True, id=None):
+        self.hyperparameters["base"] = base
+        self._id = id
+        self.queue_idx = None
+
+        self._name = f"Adjoint({self.base.name})"
+
+        if do_queue:
+            self.queue()
+
+    @property
+    def base(self):
+        return self.hyperparameters['base']
+
+    @property
+    def data(self):
+        """Trainable parameters that the operator depends on."""
+        return self.base.data
```

Comment:
@josh146  you know better, but should `data` be accessible? I thought only `parameters` should?

Reply:
I think so yes 🤔

Reply:
For one thing, we need to be able to update the data, so we need a setter that updates the base operation's data.